engine/classes/Elgg/Application.php

```diff
@@ -0,0 +1,378 @@
+
+	/**
+	 * @var string
+	 */
+	private $install_dir;
+
+	/**
+	 * @var string
+	 */
+	private $config_file;
+
+	/**
+	 * @var Application
+	 */
+	private static $test_instance;
+
+	/**
+	 * Constructor
+	 *
+	 * @param Config|string $config Config object or settings file location or
+	 */
+	public function __construct($config = null) {
+		/**
+		 * The time with microseconds when the Elgg engine was started.
+		 *
+		 * @global float
+		 */
+		global $START_MICROTIME;
+		if (!isset($START_MICROTIME)) {
+			$START_MICROTIME = microtime(true);
+		}
+
+		$this->engine_dir = dirname(dirname(__DIR__));
+		$this->install_dir = dirname($this->engine_dir);
+
+		if ($config) {
+			if (is_string($config)) {
+				$this->config_file = $config;
+			} elseif ($config instanceof Config) {
+				$this->config = $config;
+			} else {
+				throw new \InvalidArgumentException('$config must be an Elgg\Config or a settings file location');
+			}
+		} else {
+			$this->config_file = "{$this->engine_dir}/settings.php";
+		}
+	}
+
+	/**
+	 * @return Config
+	 */
+	public function getConfig() {
+		if (!$this->config) {
+			$this->loadConfig();
+		}
+		return $this->config;
+	}
+
+	/**
+	 * Load all lib files without triggering boot events
+	 *
+	 * @return void
+	 */
+	protected function loadConfig() {
+		/**
+		 * Configuration values.
+		 *
+		 * The $CONFIG global contains configuration values required
+		 * for running Elgg as defined in the settings.php file.
+		 *
+		 * Plugin authors are encouraged to use elgg_get_config() instead of accessing
+		 * the global directly.
+		 *
+		 * @see elgg_get_config()
+		 * @see engine/settings.php
+		 * @global \stdClass $CONFIG
+		 */
+		global $CONFIG;
+		if (!isset($CONFIG)) {
+			$CONFIG = new \stdClass;
+		}
+		$CONFIG->boot_complete = false;
+
+		// No settings means a fresh install
+		if (!is_file($this->config_file)) {
+			header("Location: install.php");
+			exit;
+		}
+
+		if (!is_readable($this->config_file)) {
+			echo "The Elgg settings file exists but the web server doesn't have read permission to it.";
+			exit;
+		}
+
+		require_once $this->config_file;
+
+		if (isset($CONFIG->dataroot)) {
+			$CONFIG->dataroot = rtrim($CONFIG->dataroot, DIRECTORY_SEPARATOR) . DIRECTORY_SEPARATOR;
+		}
+
+		$this->config = new Config($CONFIG);
+	}
+
+	/**
+	 * Load all Elgg code without formal boot process, for internal testing purposes
+	 *
+	 * @access private
+	 * @return void
+	 */
+	public function loadCore() {
+		if (function_exists('_elgg_services')) {
+			return;
+		}
+
+		$lib_dir = $this->engine_dir . "/lib";
+
+		// we only depend on it to be defining _elgg_services function
+		require_once "$lib_dir/autoloader.php";
+
+		// set up autoloading and DIC
+		$services = _elgg_services($this);
+
+		// load the rest of the library files from engine/lib/
+		// All on separate lines to make diffs easy to read + make it apparent how much
+		// we're actually loading on every page (Hint: it's too much).
+		$lib_files = array(
```

Comment:
I see no benefit and main gain here is to have more control over what gets loaded when.